# Instalações

In [1]:
%pip install -r requirements.txt

Note: you may need to restart the kernel to use updated packages.



[notice] A new release of pip is available: 26.0.1 -> 26.2.1
[notice] To update, run: python.exe -m pip install --upgrade pip


In [2]:
print("Dependências centralizadas em requirements.txt e instaladas na célula anterior.")

Dependências centralizadas em requirements.txt e instaladas na célula anterior.


# Pré Processamento

In [3]:
from pathlib import Path

from causal_discovery import CausalPreprocessor, load_time_series_dataset

DATASET_PROFILES = {
    "delhi_csv": {
        "data_format": "csv",
        "data_path": Path("DailyDelhiClimateTrain.csv"),
        "date_column": "date",
        "selected_columns": [
            "meantemp",
            "humidity",
            "wind_speed",
            "meanpressure",
        ],
        "decomposition_period": 30,
        "max_lag": 5,
    },
    "causaltime_traffic": {
        "data_format": "causaltime",
        "data_path": Path("datasets/causaltime/traffic/gen_data.npy"),
        "graph_path": Path("datasets/causaltime/traffic/graph.npy"),
        "trajectory_index": 0,
        "column_prefix": "traffic",
        # None usa todos os nós. A lista abaixo é um subgrafo pequeno para teste.
        "selected_columns": [
            "traffic_02",
            "traffic_04",
            "traffic_06",
            "traffic_08",
            "traffic_13",
            "traffic_19",
        ],
        "decomposition_period": None,
        "max_lag": 2,
        # Trajetorias curtas favorecem um painel parcimonioso no benchmark.
        "panel_max_lag": 1,
        # Trajetorias ja usadas como desenvolvimento/holdout em
        # Benchmark_Superioridade_Ensemble_Traffic.ipynb, mais a trajetoria 0 usada
        # acima na execucao unica: excluidas da amostra de replicas para nao reusar
        # dados ja vistos.
        "replicate_source": {
            "kind": "trajectories",
            "excluded_trajectories": [
                0,
                12, 37, 42, 84, 169, 172, 176, 207, 210, 223, 303, 307, 312, 369, 401,
                3, 40, 71, 113, 135, 170, 184, 186, 217, 230, 240, 273, 287, 298, 441,
            ],
        },
    },
}

# Troque somente este nome para alternar entre os datasets configurados.
ACTIVE_DATASET = "causaltime_traffic"
dataset_config = DATASET_PROFILES[ACTIVE_DATASET]

dataset_bundle = load_time_series_dataset(
    dataset_config["data_path"],
    data_format=dataset_config["data_format"],
    selected_columns=dataset_config.get("selected_columns"),
    date_column=dataset_config.get("date_column"),
    graph_path=dataset_config.get("graph_path"),
    trajectory_index=dataset_config.get("trajectory_index", 0),
    column_prefix=dataset_config.get("column_prefix", "variable"),
)
raw_data = dataset_bundle.data.copy()
data = raw_data.copy()
dataset_ground_truth = dataset_bundle.ground_truth.copy()
SELECTED_COLUMNS = list(dataset_bundle.selected_columns)

preprocessor = CausalPreprocessor(
    data,
    significance_level=0.05,
    decomposition_period=dataset_config.get("decomposition_period"),
)
processed_data = preprocessor.fit_transform(
    make_stationary=True,
    normalize=True,
    remove_trend=False,
    max_diffs=2,
)

max_lag = int(dataset_config.get("max_lag", 5))
preprocessing_summary = preprocessor.summary()

print(f"--- Dataset ativo: {ACTIVE_DATASET} ---")
print("Formato:", dataset_bundle.source_format)
print("Colunas disponíveis:", list(dataset_bundle.available_columns))
print("Colunas selecionadas:", SELECTED_COLUMNS)
print("Metadados:", dataset_bundle.metadata)
if not dataset_ground_truth.empty:
    print("\n--- Ground truth disponível (lag não informado) ---")
    print(dataset_ground_truth.head(20))
print("\n--- Dados originais ---")
print(data.head())
print("\n--- Resumo do pré-processamento ---")
print(preprocessing_summary)
print("\n--- Dados processados ---")
print(processed_data.head())

--- Dataset ativo: causaltime_traffic ---
Formato: causaltime
Colunas disponíveis: ['traffic_00', 'traffic_01', 'traffic_02', 'traffic_03', 'traffic_04', 'traffic_05', 'traffic_06', 'traffic_07', 'traffic_08', 'traffic_09', 'traffic_10', 'traffic_11', 'traffic_12', 'traffic_13', 'traffic_14', 'traffic_15', 'traffic_16', 'traffic_17', 'traffic_18', 'traffic_19']
Colunas selecionadas: ['traffic_02', 'traffic_04', 'traffic_06', 'traffic_08', 'traffic_13', 'traffic_19']
Metadados: {'data_path': 'datasets\\causaltime\\traffic\\gen_data.npy', 'graph_path': 'datasets\\causaltime\\traffic\\graph.npy', 'trajectory_index': 0, 'trajectory_count': 480, 'trajectory_length': 40, 'observed_node_count': 20, 'auxiliary_channel_count': 20, 'ground_truth_has_lag': False, 'ground_truth_is_symmetric': True}

--- Ground truth disponível (lag não informado) ---
        source      target   lag
0   traffic_02  traffic_04  <NA>
1   traffic_02  traffic_06  <NA>
2   traffic_02  traffic_08  <NA>
3   traffic_02  t

# Pipeline robusto

O fluxo principal usa seleção de combinações, ensemble probabilístico e estabilidade por bootstrap. Os métodos individuais são executados internamente pelo pipeline, evitando cálculos duplicados.

In [4]:
import math
import os

from causal_discovery import (
    get_registered_method_kwargs,
    get_registered_method_weights,
    get_registered_methods,
)

if "processed_data" not in globals() or "max_lag" not in globals():
    raise RuntimeError(
        "Execute primeiro a etapa de pré-processamento para gerar "
        "processed_data e max_lag."
    )

all_candidate_methods = get_registered_methods()
all_candidate_method_kwargs = get_registered_method_kwargs(max_lag)
all_method_weights = get_registered_method_weights()

# Valores iniciais reproduzíveis; a interface permite alterá-los.
quick_mode = False
expert_knowledge = []

# Limiares do segundo estágio (soft voting por consenso de precisão), usados tanto na
# execução única quanto na validação estatística replicada mais abaixo.
ENSEMBLE_THRESHOLD = 0.50
SOFT_VOTING_SUPPORT_THRESHOLD = 0.0


def runtime_method_names(quick_mode):
    return list(all_candidate_methods)


def build_selection_args(quick_mode, n_bootstrap=None, parallel_jobs=None):
    """ENSEMBLE_AUTO: com quick_mode=False (padrão), busca automaticamente todas as
    combinações de 2 a N métodos e escolhe a de maior performance_score (cego ao
    grafo verdadeiro). quick_mode=True limita a busca a 3 métodos para agilidade
    exploratória."""
    quick_mode = bool(quick_mode)
    num_methods = len(all_candidate_methods)
    default_max_methods = min(3, num_methods) if quick_mode else num_methods
    default_bootstraps = 4 if quick_mode else 8
    default_jobs = max(1, min(4, (os.cpu_count() or 2) - 1))

    return {
        "min_methods": 2,
        "max_methods": default_max_methods,
        "min_votes": 1,
        "n_bootstrap": int(n_bootstrap or default_bootstraps),
        "block_size": max(2, len(processed_data) // 12),
        "stability_threshold": 0.6,
        "selection_probability_threshold": 0.55,
        "prior_edge_probability": 0.1,
        "posterior_weight": 0.7,
        "adaptive_method_weights": True,
        "stability_weight": 0.65,
        "local_expert_weight": 0.60,
        "predictive_validation_weight": 0.75,
        "predictive_validation_max_lag": max_lag,
        "predictive_validation_splits": 3,
        "predictive_validation_ridge_alpha": 1.0,
        "method_redundancy_penalty": 0.20,
        "method_stability_power": 1.0,
        "method_diversity_bonus": 0.15,
        "method_density_penalty": 0.5,
        "minimum_method_weight": 0.05,
        "confidence_level": 0.95,
        "random_state": 42,
        "precompute_runs": True,
        "parallel_jobs": int(parallel_jobs or default_jobs),
        "max_bootstrap_seconds": 240 if quick_mode else 900,
    }


def print_selection_cost(selection_args, method_names):
    num_methods = len(method_names)
    min_methods = selection_args["min_methods"]
    max_methods = min(selection_args["max_methods"], num_methods)
    combination_count = sum(
        math.comb(num_methods, size)
        for size in range(min_methods, max_methods + 1)
    )
    optimized_runs = num_methods * (selection_args["n_bootstrap"] + 1)

    print("--- Configuração do pipeline robusto (ENSEMBLE_AUTO) ---")
    print(f"Métodos candidatos: {method_names}")
    print(f"Combinações avaliadas: {combination_count}")
    print(f"Execuções de método com pré-cálculo: {optimized_runs}")
    print(f"Paralelismo: {selection_args['parallel_jobs']} job(s)")
    print(f"Limite do bootstrap: {selection_args['max_bootstrap_seconds']} s")
    print(f"Limiar de seleção (soft voting): {ENSEMBLE_THRESHOLD:.2f}")


search_method_names = runtime_method_names(quick_mode)
common_selection_args = build_selection_args(quick_mode)
print_selection_cost(common_selection_args, search_method_names)

--- Configuração do pipeline robusto (ENSEMBLE_AUTO) ---
Métodos candidatos: ['ClassicalGranger', 'GES', 'FCI', 'DYNOTEARS', 'LPCMCI', 'NeuralGrangercMLP', 'PCMCI', 'VARLiNGAM']
Combinações avaliadas: 247
Execuções de método com pré-cálculo: 72
Paralelismo: 4 job(s)
Limite do bootstrap: 900 s
Limiar de seleção (soft voting): 0.50


# Interface interativa e visualização

In [5]:
from causal_discovery import add_precision_consensus_selection
from causal_discovery.ensemble_selection import select_robust_ensemble_combination
from causal_discovery.visualization import create_advanced_expert_dashboard


def restrict_method_relations(method, allowed_relations):
    """Mantém apenas as arestas solicitadas sem alterar os dados de entrada."""
    allowed_relations = set(allowed_relations)

    def run_restricted(data, **kwargs):
        result = method(data, **kwargs)
        if result is None or result.empty:
            return result

        selected_rows = [
            (source, target) in allowed_relations
            for source, target in zip(result["source"], result["target"])
        ]
        return result.loc[selected_rows].reset_index(drop=True)

    return run_restricted


def apply_soft_voting_precision(summary):
    """Consenso de precisão em segundo estágio (soft voting), idêntico ao usado nos
    benchmarks de superioridade do ensemble (Traffic/Toy)."""
    frame = summary.copy()
    frame["ensemble_score"] = frame["pre_validation_ensemble_score"]
    return add_precision_consensus_selection(
        frame,
        score_threshold=ENSEMBLE_THRESHOLD,
        method_support_threshold=SOFT_VOTING_SUPPORT_THRESHOLD,
    )


def pipeline_runner(
    quick_mode,
    n_bootstrap,
    parallel_jobs,
    expert_knowledge,
    selected_relations,
    processed_data,
    candidate_methods,
    candidate_method_kwargs,
    method_weights,
    analysis_objective=None,
):
    selected_relation_set = {tuple(relation) for relation in selected_relations}
    selected_names = list(candidate_methods)
    selected_methods = {
        name: restrict_method_relations(
            candidate_methods[name],
            selected_relation_set,
        )
        for name in selected_names
    }
    selected_kwargs = {
        name: candidate_method_kwargs[name]
        for name in selected_names
    }
    selected_weights = {
        name: method_weights[name]
        for name in selected_names
    }
    selection_args = build_selection_args(
        quick_mode=quick_mode,
        n_bootstrap=n_bootstrap,
        parallel_jobs=parallel_jobs,
    )

    selection_result = select_robust_ensemble_combination(
        processed_data,
        selected_methods,
        method_kwargs=selected_kwargs,
        method_weights=selected_weights,
        expert_knowledge=list(expert_knowledge),
        **selection_args,
    )
    best_evaluation = selection_result["best_evaluation"]
    raw_probabilistic_summary = best_evaluation["probabilistic_summary"]
    probabilistic_summary = apply_soft_voting_precision(raw_probabilistic_summary)

    return {
        "probabilistic_summary": probabilistic_summary,
        "raw_probabilistic_summary": raw_probabilistic_summary,
        "consistency": best_evaluation["consistency"],
        "selection_result": selection_result,
        "expert_knowledge": list(expert_knowledge),
        "selected_relations": sorted(selected_relation_set),
        "analysis_objective": dict(analysis_objective or {}),
        "selection_args": selection_args,
        "method_names": selected_names,
    }


dashboard_ui = create_advanced_expert_dashboard(
    processed_data=processed_data,
    candidate_methods=all_candidate_methods,
    candidate_method_kwargs=all_candidate_method_kwargs,
    method_weights=all_method_weights,
    all_nodes=list(processed_data.columns),
    pipeline_callback=pipeline_runner,
    initial_expert_knowledge=expert_knowledge,
    initial_quick_mode=quick_mode,
    initial_n_bootstrap=common_selection_args["n_bootstrap"],
    initial_parallel_jobs=common_selection_args["parallel_jobs"],
)

# Resultados do pipeline robusto

In [6]:
dashboard_result = getattr(dashboard_ui, "pipeline_result", None)
if not isinstance(dashboard_result, dict):
    print("--- Interface não executada: usando configuração padrão ---")
    default_relations = [
        (source, target)
        for source in processed_data.columns
        for target in processed_data.columns
        if source != target
    ]
    default_objective = {
        "type": "full_structure",
        "primary_variable": None,
        "secondary_variable": None,
    }
    dashboard_result = pipeline_runner(
        quick_mode=quick_mode,
        n_bootstrap=common_selection_args["n_bootstrap"],
        parallel_jobs=common_selection_args["parallel_jobs"],
        expert_knowledge=[],
        selected_relations=default_relations,
        processed_data=processed_data,
        candidate_methods=all_candidate_methods,
        candidate_method_kwargs=all_candidate_method_kwargs,
        method_weights=all_method_weights,
        analysis_objective=default_objective,
    )
    dashboard_ui.pipeline_result = dashboard_result
    dashboard_ui.result_summary = dashboard_result["probabilistic_summary"]
    dashboard_ui.consistency_matrix = dashboard_result["consistency"]
    dashboard_ui.selected_relations = list(default_relations)
    dashboard_ui.analysis_objective = dict(default_objective)
    execution_origin = "configuração padrão"
else:
    execution_origin = "interface interativa"

selection_result = dashboard_result["selection_result"]
ranking = selection_result["ranking"].copy()
best_eval = selection_result["best_evaluation"]
expert_knowledge = list(dashboard_result["expert_knowledge"])
selected_relations = list(dashboard_result["selected_relations"])
analysis_objective = dict(dashboard_result["analysis_objective"])
selection_args = dict(dashboard_result["selection_args"])
search_method_names = list(dashboard_result["method_names"])
probabilistic_summary = dashboard_result["probabilistic_summary"].copy()

print(f"--- Resultado do pipeline robusto ({execution_origin}) ---")
print(f"Objetivo: {analysis_objective}")
print(f"Relações analisadas: {len(selected_relations)}")
print(f"Métodos candidatos: {search_method_names}")
print(f"Melhor combinação (ENSEMBLE_AUTO): {selection_result['best_combination']}")
print("Pesos adaptativos:", best_eval["effective_method_weights"])
display(best_eval["method_weight_diagnostics"])

print("\n--- Ranking das combinações ---")
ranking_columns = [
    "combination",
    "performance_score",
    "mean_stability",
    "mean_edge_probability",
    "mean_confidence",
]
print(ranking[ranking_columns].head(10))

print("\n--- Métricas da melhor combinação ---")
print(best_eval["metrics"])

stable_edges = best_eval["stability"]
stable_edges = stable_edges.loc[stable_edges["stability_selected"]].copy()
print("\n--- Arestas estáveis ---")
print(stable_edges.head(20))

summary_columns = [
    "source",
    "target",
    "lag",
    "votes",
    "positive_votes",
    "negative_votes",
    "sign_consensus",
    "sign_agreement",
    "edge_probability",
    "ensemble_score",
    "support_ratio",
    "ensemble_selected",
    "local_expert_score",
    "consensus_score",
    "dominant_method",
    "dominant_edge_stability",
    "confidence",
    "method",
]
summary_columns = [
    column for column in summary_columns
    if column in probabilistic_summary.columns
]
print(f"\n--- Arestas da melhor combinação (ENSEMBLE_AUTO, limiar={ENSEMBLE_THRESHOLD:.2f}) ---")
print(probabilistic_summary[summary_columns].head(30))

--- Interface não executada: usando configuração padrão ---


--- Resultado do pipeline robusto (configuração padrão) ---
Objetivo: {'type': 'full_structure', 'primary_variable': None, 'secondary_variable': None}
Relações analisadas: 30
Métodos candidatos: ['ClassicalGranger', 'GES', 'FCI', 'DYNOTEARS', 'LPCMCI', 'NeuralGrangercMLP', 'PCMCI', 'VARLiNGAM']
Melhor combinação (ENSEMBLE_AUTO): ['DYNOTEARS', 'NeuralGrangercMLP']
Pesos adaptativos: {'DYNOTEARS': 0.6853350385372957, 'NeuralGrangercMLP': 1.3146649614627042}


,method,prior_weight,bootstrap_stability,diversity,mean_edge_count,relative_density,density_factor,redundancy,redundancy_factor,adaptive_weight
0,DYNOTEARS,1.0,0.516262,0.662453,55.444444,1.328895,0.858776,0.337547,0.93676,0.685335
1,NeuralGrangercMLP,1.0,0.850476,0.662453,28.000000,0.671105,1.000000,0.337547,0.93676,1.314665



--- Ranking das combinações ---
                                         combination  performance_score  \
0                      DYNOTEARS + NeuralGrangercMLP           0.703647   
1                      NeuralGrangercMLP + VARLiNGAM           0.688206   
2              DYNOTEARS + NeuralGrangercMLP + PCMCI           0.682867   
3          DYNOTEARS + NeuralGrangercMLP + VARLiNGAM           0.681513   
4                          NeuralGrangercMLP + PCMCI           0.681498   
5             DYNOTEARS + LPCMCI + NeuralGrangercMLP           0.674978   
6                            GES + NeuralGrangercMLP           0.672524   
7                GES + DYNOTEARS + NeuralGrangercMLP           0.672013   
8  DYNOTEARS + NeuralGrangercMLP + PCMCI + VARLiNGAM           0.668081   
9     DYNOTEARS + LPCMCI + NeuralGrangercMLP + PCMCI           0.667341   

   mean_stability  mean_edge_probability  mean_confidence  
0        0.861735               0.844589         0.310659  
1        0.819934    

## Validação do Ensemble contra o Grafo do Dataset

In [7]:
from causal_discovery import (
    compute_ranked_undirected_skeleton_metrics,
    compute_undirected_skeleton_metrics,
    run_pcmci_multiple_trajectories,
)

if dataset_ground_truth.empty:
    dataset_validation_metrics = None
    print("O dataset ativo não fornece ground truth para validação estrutural.")
else:
    panel_pair_scores = None
    panel_ranking_metrics = None
    if dataset_bundle.trajectory_count > 1:
        panel_max_lag = int(dataset_config.get("panel_max_lag", 1))
        panel_pair_scores = run_pcmci_multiple_trajectories(
            dataset_bundle.observed_trajectories(),
            dataset_bundle.available_columns,
            max_lag=panel_max_lag,
            standardize=True,
        )
        evaluated_pairs = {
            tuple(sorted((source, target)))
            for source, target in selected_relations
            if source != target
        }
        panel_pair_scores = panel_pair_scores.loc[[
            tuple(sorted((source, target))) in evaluated_pairs
            for source, target in zip(
                panel_pair_scores["source"],
                panel_pair_scores["target"],
            )
        ]].reset_index(drop=True)
        print("--- Evidência Complementar com Todas as Trajetórias ---")
        print(f"Trajetórias independentes: {dataset_bundle.trajectory_count}")
        print(f"Max lag do painel: {panel_max_lag}")
        print(f"Nós usados como contexto: {len(dataset_bundle.available_columns)}")
        try:
            panel_ranking_metrics = compute_ranked_undirected_skeleton_metrics(
                panel_pair_scores,
                dataset_ground_truth,
            )
            print(f"AUROC: {panel_ranking_metrics['roc_auc']:.3f}")
            print(f"Average precision: {panel_ranking_metrics['average_precision']:.3f}")
            print(f"Baseline aleatório de AP: {panel_ranking_metrics['random_average_precision']:.3f}")
        except ValueError as error:
            print(f"Ranking não calculado: {error}")
        print("Os escores não alteram o ensemble; servem como validação de ranking sem limiar.")
        display(panel_pair_scores.sort_values("score", ascending=False).head(20))
        print()

    # A seleção binária já vem do consenso de precisão (soft voting) calculado no
    # pipeline_runner; não reaplicamos um limiar sobre edge_probability aqui.
    validation_frame = probabilistic_summary
    if "ensemble_selected" in probabilistic_summary.columns:
        validation_frame = probabilistic_summary.loc[
            probabilistic_summary["ensemble_selected"].fillna(False).astype(bool)
        ].copy()

    dataset_validation_metrics = compute_undirected_skeleton_metrics(
        validation_frame,
        dataset_ground_truth,
        prob_threshold=0.0,
        nodes=SELECTED_COLUMNS,
        evaluated_relations=selected_relations,
    )

    print("--- Validação do Esqueleto Não Direcionado (ENSEMBLE_AUTO) ---")
    print(f"Limiar de seleção (soft voting): {ENSEMBLE_THRESHOLD:.2f}")
    print(f"Pares possíveis: {dataset_validation_metrics['candidate_pairs']}")
    print(f"Pares presentes no ground truth: {dataset_validation_metrics['ground_truth_pairs']}")
    print(f"Prevalência estrutural: {dataset_validation_metrics['ground_truth_prevalence']:.2%}")

    for metric_name in [
        "true_positives",
        "false_positives",
        "false_negatives",
        "precision",
        "recall",
        "f1_score",
        "structural_hamming_distance",
    ]:
        metric_value = dataset_validation_metrics[metric_name]
        formatted_value = f"{metric_value:.2f}" if isinstance(metric_value, float) else metric_value
        print(f"{metric_name}: {formatted_value}")

    baseline_f1 = dataset_validation_metrics["all_pairs_baseline_f1"]
    f1_gain = dataset_validation_metrics["f1_score"] - baseline_f1
    print(f"F1 do baseline que prevê todos os pares: {baseline_f1:.2f}")
    print(f"Diferença de F1 do ensemble para o baseline: {f1_gain:+.2f}")

    if dataset_validation_metrics["ground_truth_prevalence"] >= 0.8:
        print("Aviso: o subconjunto selecionado é muito denso; use todos os nós para uma validação discriminativa.")

    print("Pares corretos:", dataset_validation_metrics["true_positive_pairs"])
    print("Pares extras:", dataset_validation_metrics["false_positive_pairs"])
    print("Pares não recuperados:", dataset_validation_metrics["false_negative_pairs"])
    print("Observação: direção e lag são ignorados porque o graph.npy não contém essas informações.")

--- Evidência Complementar com Todas as Trajetórias ---
Trajetórias independentes: 480
Max lag do painel: 1
Nós usados como contexto: 20
AUROC: 1.000
Average precision: 1.000
Baseline aleatório de AP: 0.933
Os escores não alteram o ensemble; servem como validação de ranking sem limiar.


,source,target,score,p_value
5,traffic_04,traffic_06,0.131248,8.105244e-71
11,traffic_06,traffic_19,0.112540,1.948117e-52
8,traffic_04,traffic_19,0.096521,5.658244e-39
3,traffic_02,traffic_13,0.063322,1.163772e-17
10,traffic_06,traffic_13,0.048106,8.179945e-11
1,traffic_02,traffic_06,0.045773,6.336740e-10
4,traffic_02,traffic_19,0.039670,8.457287e-08
0,traffic_02,traffic_04,0.037415,4.367066e-07
7,traffic_04,traffic_13,0.034948,2.371308e-06
9,traffic_06,traffic_08,0.023961,1.217369e-03



--- Validação do Esqueleto Não Direcionado (ENSEMBLE_AUTO) ---
Limiar de seleção (soft voting): 0.50
Pares possíveis: 15
Pares presentes no ground truth: 14
Prevalência estrutural: 93.33%
true_positives: 12
false_positives: 1
false_negatives: 2
precision: 0.92
recall: 0.86
f1_score: 0.89
structural_hamming_distance: 3
F1 do baseline que prevê todos os pares: 0.97
Diferença de F1 do ensemble para o baseline: -0.08
Aviso: o subconjunto selecionado é muito denso; use todos os nós para uma validação discriminativa.
Pares corretos: [('traffic_02', 'traffic_04'), ('traffic_02', 'traffic_06'), ('traffic_02', 'traffic_08'), ('traffic_02', 'traffic_13'), ('traffic_02', 'traffic_19'), ('traffic_04', 'traffic_06'), ('traffic_04', 'traffic_08'), ('traffic_04', 'traffic_13'), ('traffic_06', 'traffic_13'), ('traffic_06', 'traffic_19'), ('traffic_08', 'traffic_19'), ('traffic_13', 'traffic_19')]
Pares extras: [('traffic_08', 'traffic_13')]
Pares não recuperados: [('traffic_04', 'traffic_19'), ('traf

## Comparação: Ensemble versus algoritmos avulsos

Executa a mesma avaliação estrutural da seção anterior, mas estratégia por estratégia: cada
algoritmo registrado é avaliado sozinho e o `ENSEMBLE` é avaliado com o limiar
`dataset_validation_threshold` sobre `edge_probability`.

As saídas individuais são reaproveitadas de `selection_result["precomputed_outputs"]`, o
resultado-base que a seleção robusta já calculou uma única vez para cada método — nenhum
algoritmo é reexecutado aqui.

Convenções, idênticas às da validação anterior: direção e lag são colapsados no mesmo par de
nós; para os métodos avulsos toda aresta retornada conta como detecção; para o ensemble apenas
as arestas acima do limiar. `average_precision` e `roc_auc` avaliam o ranking completo dos pares
sem limiar, usando `1 - p_value` quando o método reporta significância e a magnitude do `score`
caso contrário.

Esta seção apenas executa e tabula. Nenhuma conclusão é derivada automaticamente.

In [8]:
# Comparação direta entre o ensemble (ENSEMBLE_AUTO: busca automática 2..N métodos
# + soft voting por consenso de precisão) e cada algoritmo executado sozinho.
import numpy as np
import pandas as pd
import plotly.express as px

from causal_discovery import (
    build_complete_undirected_pair_scores,
    compute_ranked_undirected_skeleton_metrics,
    compute_undirected_skeleton_metrics,
)

if "selection_result" not in globals() or "probabilistic_summary" not in globals():
    raise RuntimeError(
        "Execute primeiro as células do pipeline robusto e da validação estrutural."
    )


def drop_self_links(frame):
    if frame is None or len(frame) == 0:
        return pd.DataFrame(columns=["source", "target", "lag"])
    return frame.loc[frame["source"].ne(frame["target"])].reset_index(drop=True)


def ranking_evidence(frame):
    """Usa 1 - p_value quando o método reporta significância; senão, a magnitude do score."""
    if "p_value" in frame.columns:
        p_values = pd.to_numeric(frame["p_value"], errors="coerce")
        if np.isfinite(p_values).any():
            return "one_minus_p_value"
    return "absolute_score"


def undirected_pairs(frame):
    if len(frame) == 0:
        return set()
    return {
        tuple(sorted((str(source), str(target))))
        for source, target in zip(frame["source"], frame["target"])
        if source != target
    }


def format_pairs(pairs):
    """Compacta a lista de pares em uma string legível na tabela."""
    return ", ".join(f"{source}~{target}" for source, target in sorted(pairs)) or "-"


base_outputs = selection_result.get("precomputed_outputs") or {}
if not base_outputs:
    # Sem pré-cálculo, cada método ainda aparece em alguma combinação avaliada.
    base_outputs = {}
    for evaluation in selection_result["all_evaluations"].values():
        for method_name, method_output in evaluation["outputs"].items():
            base_outputs.setdefault(method_name, method_output)

individual_outputs = {
    name: drop_self_links(base_outputs[name]) for name in sorted(base_outputs)
}
ensemble_summary = drop_self_links(probabilistic_summary)
if "ensemble_selected" in ensemble_summary.columns:
    ensemble_detected = ensemble_summary.loc[
        ensemble_summary["ensemble_selected"].fillna(False).astype(bool)
    ].reset_index(drop=True)
else:
    ensemble_detected = ensemble_summary.loc[
        pd.to_numeric(ensemble_summary["edge_probability"], errors="coerce")
        >= ENSEMBLE_THRESHOLD
    ].reset_index(drop=True)

evaluated_pairs = {
    tuple(sorted((source, target)))
    for source, target in selected_relations
    if source != target
}

strategy_frames = dict(individual_outputs)
strategy_frames["ENSEMBLE_AUTO"] = ensemble_summary
strategy_evidence = {
    name: ranking_evidence(frame) for name, frame in individual_outputs.items()
}
strategy_evidence["ENSEMBLE_AUTO"] = "ensemble_score"

comparison_rows = []
detected_pairs_by_strategy = {}
pair_diagnostics = {}

for strategy, frame in strategy_frames.items():
    binary_frame = ensemble_detected if strategy == "ENSEMBLE_AUTO" else frame
    pairs = undirected_pairs(binary_frame) & evaluated_pairs
    detected_pairs_by_strategy[strategy] = pairs
    row = {
        "strategy": strategy,
        "arestas_retornadas": int(len(binary_frame)),
        "pares_detectados": len(pairs),
    }

    if not dataset_ground_truth.empty:
        binary = compute_undirected_skeleton_metrics(
            binary_frame,
            dataset_ground_truth,
            prob_threshold=0.0,
            nodes=SELECTED_COLUMNS,
            evaluated_relations=selected_relations,
        )
        pair_scores = build_complete_undirected_pair_scores(
            frame,
            SELECTED_COLUMNS,
            evidence=strategy_evidence[strategy],
        )
        pair_scores = pair_scores.loc[[
            tuple(sorted((source, target))) in evaluated_pairs
            for source, target in zip(pair_scores["source"], pair_scores["target"])
        ]].reset_index(drop=True)
        try:
            ranked = compute_ranked_undirected_skeleton_metrics(
                pair_scores, dataset_ground_truth
            )
        except ValueError as error:
            print(f"Ranking não calculado para {strategy}: {error}")
            ranked = {"average_precision": float("nan"), "roc_auc": float("nan")}

        row.update({
            "precision": binary["precision"],
            "recall": binary["recall"],
            "f1_score": binary["f1_score"],
            "f1_menos_baseline": binary["f1_score"] - binary["all_pairs_baseline_f1"],
            "structural_hamming_distance": binary["structural_hamming_distance"],
            "true_positives": binary["true_positives"],
            "false_positives": binary["false_positives"],
            "false_negatives": binary["false_negatives"],
            "average_precision": ranked["average_precision"],
            "roc_auc": ranked["roc_auc"],
        })
        pair_diagnostics[strategy] = {
            "pares_extras_fp": format_pairs(binary["false_positive_pairs"]),
            "pares_nao_recuperados_fn": format_pairs(binary["false_negative_pairs"]),
        }

    comparison_rows.append(row)

method_comparison = pd.DataFrame(comparison_rows).set_index("strategy")
if "f1_score" in method_comparison.columns:
    method_comparison = method_comparison.sort_values("f1_score", ascending=False)

print("--- Ensemble versus algoritmos avulsos ---")
print(f"Dataset ativo: {ACTIVE_DATASET}")
print(f"Melhor combinação do ensemble (ENSEMBLE_AUTO): {selection_result['best_combination']}")
print(f"Limiar de seleção (soft voting): {ENSEMBLE_THRESHOLD:.2f}")
print(f"Estratégias avaliadas: {len(method_comparison)}")
print(f"Pares avaliados: {len(evaluated_pairs)}")

if dataset_ground_truth.empty:
    print("Sem ground truth no dataset ativo: apenas contagens e sobreposição entre estratégias.")
else:
    truth_reference = compute_undirected_skeleton_metrics(
        pd.DataFrame(columns=["source", "target", "lag"]),
        dataset_ground_truth,
        nodes=SELECTED_COLUMNS,
        evaluated_relations=selected_relations,
    )
    print(f"Pares no ground truth: {truth_reference['ground_truth_pairs']}")
    print(f"Prevalência estrutural: {truth_reference['ground_truth_prevalence']:.2%}")
    print(f"F1 do baseline que prevê todos os pares: {truth_reference['all_pairs_baseline_f1']:.3f}")

display(method_comparison.round(3))

if pair_diagnostics:
    print("\n--- Erros por estratégia (pares corretos = verdadeiros menos os FN) ---")
    display(
        pd.DataFrame(pair_diagnostics).T.reindex(method_comparison.index)
    )

ensemble_pairs = detected_pairs_by_strategy["ENSEMBLE_AUTO"]
overlap_rows = []
for strategy, pairs in detected_pairs_by_strategy.items():
    if strategy == "ENSEMBLE_AUTO":
        continue
    union = pairs | ensemble_pairs
    overlap_rows.append({
        "strategy": strategy,
        "pares_em_comum": len(pairs & ensemble_pairs),
        "so_no_metodo": len(pairs - ensemble_pairs),
        "so_no_ensemble": len(ensemble_pairs - pairs),
        "jaccard": (len(pairs & ensemble_pairs) / len(union)) if union else float("nan"),
    })

print("\n--- Sobreposição de cada método com o ensemble (pares não direcionados) ---")
display(pd.DataFrame(overlap_rows).set_index("strategy").round(3))

if "f1_score" in method_comparison.columns:
    chart_frame = (
        method_comparison[["precision", "recall", "f1_score"]]
        .reset_index()
        .melt(id_vars="strategy", var_name="métrica", value_name="valor")
    )
    figure = px.bar(
        chart_frame,
        x="strategy",
        y="valor",
        color="métrica",
        barmode="group",
        category_orders={"strategy": list(method_comparison.index)},
        title=f"Ensemble (auto) versus algoritmos avulsos - {ACTIVE_DATASET}",
    )
    figure.update_xaxes(tickangle=45)
    figure.update_yaxes(range=[0, 1])
    figure.show()

--- Ensemble versus algoritmos avulsos ---
Dataset ativo: causaltime_traffic
Melhor combinação do ensemble (ENSEMBLE_AUTO): ['DYNOTEARS', 'NeuralGrangercMLP']
Limiar de seleção (soft voting): 0.50
Estratégias avaliadas: 9
Pares avaliados: 15
Pares no ground truth: 14
Prevalência estrutural: 93.33%
F1 do baseline que prevê todos os pares: 0.966


,arestas_retornadas,pares_detectados,precision,recall,f1_score,f1_menos_baseline,structural_hamming_distance,true_positives,false_positives,false_negatives,average_precision,roc_auc
strategy,,,,,,,,,,,,
DYNOTEARS,53,15,0.933,1.000,0.966,0.000,1,14,1,0,1.000,1.000
VARLiNGAM,47,15,0.933,1.000,0.966,0.000,1,14,1,0,1.000,1.000
NeuralGrangercMLP,29,15,0.933,1.000,0.966,0.000,1,14,1,0,0.957,0.500
ENSEMBLE_AUTO,16,13,0.923,0.857,0.889,-0.077,3,12,1,2,0.965,0.571
PCMCI,16,10,0.900,0.643,0.750,-0.216,6,9,1,5,0.976,0.643
ClassicalGranger,12,7,1.000,0.500,0.667,-0.299,7,7,0,7,0.967,0.750
GES,10,8,0.875,0.500,0.636,-0.329,8,7,1,7,0.904,0.250
FCI,2,2,1.000,0.143,0.250,-0.716,12,2,0,12,0.943,0.571
LPCMCI,1,1,1.000,0.071,0.133,-0.832,13,1,0,13,0.938,0.536



--- Erros por estratégia (pares corretos = verdadeiros menos os FN) ---


,pares_extras_fp,pares_nao_recuperados_fn
strategy,,
DYNOTEARS,traffic_08~traffic_13,-
VARLiNGAM,traffic_08~traffic_13,-
NeuralGrangercMLP,traffic_08~traffic_13,-
ENSEMBLE_AUTO,traffic_08~traffic_13,"traffic_04~traffic_19, traffic_06~traffic_08"
PCMCI,traffic_08~traffic_13,"traffic_02~traffic_04, traffic_02~traffic_06, ..."
ClassicalGranger,-,"traffic_02~traffic_04, traffic_02~traffic_08, ..."
GES,traffic_08~traffic_13,"traffic_02~traffic_04, traffic_02~traffic_06, ..."
FCI,-,"traffic_02~traffic_04, traffic_02~traffic_06, ..."
LPCMCI,-,"traffic_02~traffic_04, traffic_02~traffic_06, ..."



--- Sobreposição de cada método com o ensemble (pares não direcionados) ---


,pares_em_comum,so_no_metodo,so_no_ensemble,jaccard
strategy,,,,
ClassicalGranger,5,2,8,0.333
DYNOTEARS,13,2,0,0.867
FCI,2,0,11,0.154
GES,7,1,6,0.500
LPCMCI,1,0,12,0.077
NeuralGrangercMLP,13,2,0,0.867
PCMCI,9,1,4,0.643
VARLiNGAM,13,2,0,0.867


## Validação estatística com réplicas independentes (opcional, desativada por padrão)

A comparação acima é uma única execução — não é evidência estatística de superioridade. Esta
seção, desligada por padrão (`RUN_REPLICATED_VALIDATION = False`), reexecuta o mesmo pipeline
(`ENSEMBLE_AUTO`: busca automática de 2 a N métodos + soft voting por consenso de precisão)
sobre várias réplicas/trajetórias independentes do dataset ativo, e aplica o mesmo protocolo
pareado (Wilcoxon + correção de Holm + IC 95% + taxa de vitórias, `MIN_CONFIRMATORY_REPLICATES`
réplicas mínimas) usado em `Benchmark_Superioridade_Ensemble_Traffic.ipynb` e
`Benchmark_Superioridade_Ensemble_Toy.ipynb`.

Só funciona com perfis de dataset que exponham réplicas/trajetórias de verdade
(chave `"replicate_source"` em `DATASET_PROFILES`); hoje isso é só `causaltime_traffic`
(480 trajetórias — as já usadas como desenvolvimento/holdout nos benchmarks dedicados, mais a
trajetória 0 já usada na execução única acima, são excluídas da amostra). `delhi_csv` é uma
única série contínua sem réplicas: ativar a flag com esse perfil ativo levanta um erro explicando
o que falta.

**Custo:** a busca completa (2 a 8 métodos, 247 combinações) leva cerca de 5 minutos por réplica
independentemente do tamanho da série (o custo é da busca combinatória, não dos dados). Com
`N_REPLICATES=10` (o padrão), a execução completa leva cerca de 50 minutos.

Mude `RUN_REPLICATED_VALIDATION` para `True` na célula abaixo para executar. Sem conclusão
textual automática — apenas execução e tabelas, para você interpretar.

In [9]:
import numpy as np

RUN_REPLICATED_VALIDATION = False  # mude para True para rodar a validação com réplicas
N_REPLICATES = 10
REPLICATE_SAMPLE_SEED = 2029
STATISTICAL_BOOTSTRAPS = 10_000
SIGNIFICANCE_LEVEL = 0.05
MINIMUM_PRECISION_GAIN = 0.05
MINIMUM_WIN_RATE = 0.70
MIN_CONFIRMATORY_REPLICATES = 10
REPLICATED_RESULTS_DIR = Path(".local/results/series_temporais_replicated_validation") / ACTIVE_DATASET
RETRY_FAILURES = True

if not RUN_REPLICATED_VALIDATION:
    print("RUN_REPLICATED_VALIDATION=False: seção desativada. Mude a flag acima para True para executar.")
else:
    replicate_source = dataset_config.get("replicate_source")
    if replicate_source is None:
        raise RuntimeError(
            f"O perfil '{ACTIVE_DATASET}' não define 'replicate_source' (nenhuma réplica "
            "independente disponível). Troque ACTIVE_DATASET para um perfil com réplicas "
            "(ex.: 'causaltime_traffic') ou adicione 'replicate_source' ao perfil ativo."
        )
    if dataset_ground_truth.empty:
        raise RuntimeError(
            "O dataset ativo não tem ground truth: a validação pareada de precisão exige um "
            "grafo de referência conhecido."
        )
    if replicate_source["kind"] == "trajectories":
        excluded = {int(index) for index in replicate_source.get("excluded_trajectories", [])}
        available_indices = np.setdiff1d(
            np.arange(dataset_bundle.trajectory_count),
            np.asarray(sorted(excluded), dtype=int),
        )
        replicate_rng = np.random.default_rng(REPLICATE_SAMPLE_SEED)
        replicate_ids = sorted(replicate_rng.choice(
            available_indices,
            size=min(N_REPLICATES, len(available_indices)),
            replace=False,
        ).tolist())
    else:
        raise RuntimeError(f"Tipo de replicate_source não suportado: {replicate_source['kind']!r}")
    print(f"Réplicas selecionadas ({len(replicate_ids)}): {replicate_ids}")
    print(f"Tempo estimado (~5 min/réplica com busca completa): {len(replicate_ids) * 5} min")

RUN_REPLICATED_VALIDATION=False: seção desativada. Mude a flag acima para True para executar.


In [10]:
if RUN_REPLICATED_VALIDATION:
    import json
    import time

    from causal_discovery import (
        build_complete_undirected_pair_scores,
        compute_paired_superiority_statistics,
        compute_ranked_undirected_skeleton_metrics,
        compute_undirected_skeleton_metrics,
    )

    REPLICATED_METHOD_NAMES = list(all_candidate_methods.keys())
    FULL_COMBINATION_KEY = " + ".join(REPLICATED_METHOD_NAMES)

    def load_replicate_data(replicate_id):
        raw = dataset_bundle.trajectory_frame(int(replicate_id))
        preprocessor = CausalPreprocessor(
            raw, significance_level=0.05,
            decomposition_period=dataset_config.get("decomposition_period"),
        )
        return preprocessor.fit_transform(
            make_stationary=True, normalize=True, remove_trend=False, max_diffs=2,
        )

    def replicate_evidence_mode(frame):
        if "p_value" in frame.columns:
            p_values = pd.to_numeric(frame["p_value"], errors="coerce")
            if np.isfinite(p_values).any():
                return "one_minus_p_value"
        return "absolute_score"

    def evaluate_replicate_strategy(frame, strategy, replicate_id, runtime_seconds, *, probability=False):
        binary_frame = frame
        if probability and "ensemble_selected" in frame:
            binary_frame = frame.loc[
                frame["ensemble_selected"].fillna(False).astype(bool)
            ].copy()
        binary = compute_undirected_skeleton_metrics(
            binary_frame, dataset_ground_truth, prob_threshold=0.0,
            nodes=SELECTED_COLUMNS, evaluated_relations=selected_relations,
        )
        pair_scores = build_complete_undirected_pair_scores(
            frame, SELECTED_COLUMNS,
            evidence=(
                "ensemble_score" if probability and "ensemble_score" in frame.columns
                else "probability" if probability
                else replicate_evidence_mode(frame)
            ),
        )
        try:
            ranked = compute_ranked_undirected_skeleton_metrics(pair_scores, dataset_ground_truth)
        except ValueError:
            ranked = {"average_precision": float("nan"), "roc_auc": float("nan")}
        return {
            "replicate_id": int(replicate_id), "strategy": str(strategy),
            "precision": binary["precision"], "recall": binary["recall"],
            "f1_score": binary["f1_score"],
            "structural_hamming_distance": binary["structural_hamming_distance"],
            "true_positives": binary["true_positives"],
            "false_positives": binary["false_positives"],
            "false_negatives": binary["false_negatives"],
            "average_precision": ranked["average_precision"], "roc_auc": ranked["roc_auc"],
            "runtime_seconds": float(runtime_seconds),
        }

    def replicate_baseline_rows(replicate_id):
        pairs = [
            (SELECTED_COLUMNS[i], SELECTED_COLUMNS[j])
            for i in range(len(SELECTED_COLUMNS)) for j in range(i + 1, len(SELECTED_COLUMNS))
        ]
        all_pairs = pd.DataFrame([
            {"source": s, "target": t, "lag": 1, "score": 1.0, "p_value": np.nan} for s, t in pairs
        ])
        rng = np.random.default_rng(REPLICATE_SAMPLE_SEED + int(replicate_id))
        random_scores = rng.random(len(pairs))
        true_pair_count = compute_undirected_skeleton_metrics(
            pd.DataFrame(columns=["source", "target", "lag"]), dataset_ground_truth, nodes=SELECTED_COLUMNS,
        )["ground_truth_pairs"]
        selected = np.argsort(random_scores)[-true_pair_count:] if true_pair_count else np.array([], dtype=int)
        random_edges = pd.DataFrame([
            {"source": pairs[i][0], "target": pairs[i][1], "lag": 1, "score": random_scores[i], "p_value": np.nan}
            for i in selected
        ])
        random_pair_scores = pd.DataFrame([
            {"source": s, "target": t, "score": score} for (s, t), score in zip(pairs, random_scores)
        ])
        all_metrics = evaluate_replicate_strategy(all_pairs, "ALL_PAIRS", replicate_id, 0.0)
        random_binary = compute_undirected_skeleton_metrics(random_edges, dataset_ground_truth, nodes=SELECTED_COLUMNS)
        random_ranked = compute_ranked_undirected_skeleton_metrics(random_pair_scores, dataset_ground_truth)
        random_metrics = {
            "replicate_id": int(replicate_id), "strategy": "RANDOM_DENSITY",
            "precision": random_binary["precision"], "recall": random_binary["recall"],
            "f1_score": random_binary["f1_score"],
            "structural_hamming_distance": random_binary["structural_hamming_distance"],
            "true_positives": random_binary["true_positives"], "false_positives": random_binary["false_positives"],
            "false_negatives": random_binary["false_negatives"],
            "average_precision": random_ranked["average_precision"], "roc_auc": random_ranked["roc_auc"],
            "runtime_seconds": 0.0,
        }
        return [all_metrics, random_metrics]

    def run_replicate(replicate_id):
        processed = load_replicate_data(replicate_id)
        relations = {(s, t) for s in processed.columns for t in processed.columns if s != t}
        restricted_methods = {
            name: restrict_method_relations(all_candidate_methods[name], relations)
            for name in REPLICATED_METHOD_NAMES
        }
        selection = select_robust_ensemble_combination(
            processed, restricted_methods,
            method_kwargs=all_candidate_method_kwargs, method_weights=all_method_weights,
            expert_knowledge=[], **build_selection_args(quick_mode=False),
        )
        base_outputs = selection.get("precomputed_outputs") or {}
        if not base_outputs:
            base_outputs = {}
            for evaluation in selection["all_evaluations"].values():
                for name, output in evaluation["outputs"].items():
                    base_outputs.setdefault(name, output)

        rows = [
            evaluate_replicate_strategy(
                base_outputs[name].loc[base_outputs[name]["source"].ne(base_outputs[name]["target"])],
                name, replicate_id, np.nan,
            )
            for name in REPLICATED_METHOD_NAMES
        ]

        auto_summary = apply_soft_voting_precision(selection["best_evaluation"]["probabilistic_summary"].copy())
        rows.append(evaluate_replicate_strategy(auto_summary, "ENSEMBLE_AUTO", replicate_id, 0.0, probability=True))

        full_evaluation = selection["all_evaluations"][FULL_COMBINATION_KEY]
        full_summary = apply_soft_voting_precision(full_evaluation["probabilistic_summary"].copy())
        rows.append(evaluate_replicate_strategy(full_summary, "ENSEMBLE", replicate_id, 0.0, probability=True))

        rows.extend(replicate_baseline_rows(replicate_id))

        selection_row = {
            "replicate_id": int(replicate_id),
            "auto_combination": " + ".join(selection["best_combination"]),
            "combinations_evaluated": int(len(selection["ranking"])),
        }
        return rows, selection_row

    REPLICATED_RESULTS_DIR.mkdir(parents=True, exist_ok=True)
    replicated_metrics_path = REPLICATED_RESULTS_DIR / "metrics.csv"
    replicated_selections_path = REPLICATED_RESULTS_DIR / "selections.csv"
    replicated_failures_path = REPLICATED_RESULTS_DIR / "failures.csv"

    replicated_metrics_results = (
        pd.read_csv(replicated_metrics_path) if replicated_metrics_path.exists() else pd.DataFrame()
    )
    replicated_selection_results = (
        pd.read_csv(replicated_selections_path) if replicated_selections_path.exists() else pd.DataFrame()
    )
    replicated_failure_results = (
        pd.read_csv(replicated_failures_path) if replicated_failures_path.exists() else pd.DataFrame()
    )
    completed_replicates = set(
        replicated_metrics_results.loc[
            replicated_metrics_results.get("strategy", pd.Series(dtype=str)).eq("ENSEMBLE_AUTO"),
            "replicate_id",
        ].astype(int)
    ) if not replicated_metrics_results.empty else set()
    failed_replicates = (
        set(replicated_failure_results["replicate_id"].astype(int))
        if not replicated_failure_results.empty else set()
    )

    for position, replicate_id in enumerate(replicate_ids, start=1):
        if replicate_id in completed_replicates or (replicate_id in failed_replicates and not RETRY_FAILURES):
            print(f"[{position}/{len(replicate_ids)}] réplica {replicate_id}: checkpoint")
            continue
        print(f"[{position}/{len(replicate_ids)}] réplica {replicate_id}: executando...")
        started = time.perf_counter()
        try:
            rows, selection_row = run_replicate(replicate_id)
            replicated_metrics_results = pd.concat(
                [replicated_metrics_results, pd.DataFrame(rows)], ignore_index=True
            )
            replicated_selection_results = pd.concat(
                [replicated_selection_results, pd.DataFrame([selection_row])], ignore_index=True
            )
            replicated_metrics_results.to_csv(replicated_metrics_path, index=False)
            replicated_selection_results.to_csv(replicated_selections_path, index=False)
            if not replicated_failure_results.empty:
                replicated_failure_results = replicated_failure_results.loc[
                    replicated_failure_results["replicate_id"].astype(int).ne(int(replicate_id))
                ].reset_index(drop=True)
                if replicated_failure_results.empty:
                    replicated_failures_path.unlink(missing_ok=True)
                else:
                    replicated_failure_results.to_csv(replicated_failures_path, index=False)
            print(f"  concluída em {(time.perf_counter() - started) / 60:.1f} min")
        except Exception as error:
            failure_row = {
                "replicate_id": int(replicate_id),
                "error_type": type(error).__name__, "error": str(error),
            }
            replicated_failure_results = pd.concat(
                [replicated_failure_results, pd.DataFrame([failure_row])], ignore_index=True
            )
            replicated_failure_results.to_csv(replicated_failures_path, index=False)
            print(f"  FALHA: {type(error).__name__}: {error}")

    completed_count = (
        replicated_metrics_results.loc[
            replicated_metrics_results["strategy"].eq("ENSEMBLE_AUTO"), "replicate_id"
        ].nunique() if not replicated_metrics_results.empty else 0
    )
    print(f"Réplicas completas: {completed_count}")
    print(f"Falhas registradas: {len(replicated_failure_results)}")
else:
    print("RUN_REPLICATED_VALIDATION=False: execução pulada.")

RUN_REPLICATED_VALIDATION=False: execução pulada.


In [11]:
if RUN_REPLICATED_VALIDATION:
    replicated_metrics_results = (
        pd.read_csv(replicated_metrics_path) if replicated_metrics_path.exists() else pd.DataFrame()
    )
    if replicated_metrics_results.empty:
        raise RuntimeError("Nenhum resultado disponível. Execute a célula de execução acima.")

    replicated_main_strategies = [*REPLICATED_METHOD_NAMES, "ENSEMBLE", "ENSEMBLE_AUTO"]
    replicated_strategy_order = ["ENSEMBLE_AUTO", "ENSEMBLE", *sorted(REPLICATED_METHOD_NAMES)]
    replicated_display = replicated_metrics_results.loc[
        replicated_metrics_results["strategy"].isin(replicated_main_strategies)
    ].copy()
    replicated_descriptive = replicated_display.groupby("strategy")[[
        "precision", "recall", "f1_score", "average_precision", "roc_auc",
    ]].agg(["mean", "std"]).round(4).reindex(replicated_strategy_order)
    print(f"--- {ACTIVE_DATASET}: média e desvio-padrão entre {len(replicate_ids)} réplicas ---")
    display(replicated_descriptive)

    replicated_figure = px.box(
        replicated_display, x="strategy", y="precision", points="all",
        category_orders={"strategy": replicated_strategy_order},
        title=f"Precisão por estratégia - {ACTIVE_DATASET} ({len(replicate_ids)} réplicas)",
    )
    replicated_figure.update_xaxes(tickangle=45)
    replicated_figure.show()

    print("Subconjuntos escolhidos pela busca automática nas réplicas:")
    print(replicated_selection_results["auto_combination"].value_counts().to_string())
else:
    print("RUN_REPLICATED_VALIDATION=False: seção estatística pulada.")

RUN_REPLICATED_VALIDATION=False: seção estatística pulada.


In [12]:
if RUN_REPLICATED_VALIDATION:
    def holm_adjust(p_values):
        values = np.asarray(p_values, dtype=float)
        order = np.argsort(values)
        adjusted_sorted = np.maximum.accumulate(
            np.array([(len(values) - rank) * values[index] for rank, index in enumerate(order)])
        )
        adjusted = np.empty_like(values)
        adjusted[order] = np.minimum(adjusted_sorted, 1.0)
        return adjusted

    replicated_baselines = [*REPLICATED_METHOD_NAMES, "ENSEMBLE"]
    replicated_stats_frame = replicated_metrics_results.rename(columns={"replicate_id": "trajectory_index"})
    replicated_primary_rows = [
        compute_paired_superiority_statistics(
            replicated_stats_frame, candidate="ENSEMBLE_AUTO", baseline=baseline,
            metric="precision", higher_is_better=True,
            n_bootstrap=STATISTICAL_BOOTSTRAPS, random_state=42,
        )
        for baseline in replicated_baselines
    ]
    replicated_primary_comparison = pd.DataFrame(replicated_primary_rows)
    replicated_primary_comparison["holm_p_value"] = holm_adjust(replicated_primary_comparison["wilcoxon_p_value"])
    replicated_primary_comparison["confirmatory_sample_available"] = (
        replicated_primary_comparison["paired_trajectories"] >= MIN_CONFIRMATORY_REPLICATES
    )
    replicated_primary_comparison["superiority_criterion_met"] = (
        replicated_primary_comparison["confirmatory_sample_available"]
    ) & (
        replicated_primary_comparison["confidence_interval_low"] > MINIMUM_PRECISION_GAIN
    ) & (
        replicated_primary_comparison["holm_p_value"] < SIGNIFICANCE_LEVEL
    ) & (
        replicated_primary_comparison["win_rate"] >= MINIMUM_WIN_RATE
    )
    display(replicated_primary_comparison.sort_values("baseline").round(4))
else:
    print("RUN_REPLICATED_VALIDATION=False: estatística pulada.")

RUN_REPLICATED_VALIDATION=False: estatística pulada.


# Benchmark e Validação com Ground Truth

In [13]:
from causal_discovery.benchmark import generate_synthetic_timeseries, compute_structural_metrics

# 1. Gera Dataset Controlado
df_synth, ground_truth = generate_synthetic_timeseries(n_samples=500)
ground_truth = ground_truth.loc[
    ground_truth["source"].ne(ground_truth["target"])
].reset_index(drop=True)

print("--- Ground Truth Esperado ---")
print(ground_truth)

# 2. Executa Ensemble no sintético
from causal_discovery.ensemble_selection import select_robust_ensemble_combination

benchmark_methods = dict(all_candidate_methods)
benchmark_method_kwargs = {
    name: {**kwargs, "max_lag": 2}
    for name, kwargs in all_candidate_method_kwargs.items()
}
# Mantém o Neural Granger mais leve apenas no benchmark auxiliar.
benchmark_method_kwargs["NeuralGrangercMLP"]["max_iter"] = 200

# Configuração robusta do benchmark: consenso de 2 entre 3 métodos.
benchmark_min_methods = 3
benchmark_max_methods = 3
benchmark_min_votes = 2
benchmark_n_bootstrap = 20
benchmark_selection_threshold = 0.6
benchmark_probability_threshold = 0.5

synthetic_selection = select_robust_ensemble_combination(
    df_synth,
    benchmark_methods,
    method_kwargs=benchmark_method_kwargs,
    method_weights=all_method_weights,
    min_methods=benchmark_min_methods,
    max_methods=benchmark_max_methods,
    min_votes=benchmark_min_votes,
    n_bootstrap=benchmark_n_bootstrap,
    selection_probability_threshold=benchmark_selection_threshold,
    random_state=42,
)

synth_summary = synthetic_selection["best_evaluation"]["probabilistic_summary"]
synth_summary = synth_summary.loc[
    synth_summary["source"].ne(synth_summary["target"])
].reset_index(drop=True)

# 3. Calcula as Métricas SHD, Precision, F1-Score
metrics = compute_structural_metrics(
    synth_summary,
    ground_truth,
    prob_threshold=benchmark_probability_threshold,
)

edge_columns = ["source", "target", "lag"]
predicted_edges = set(map(
    tuple,
    synth_summary.loc[
        synth_summary["edge_probability"] >= benchmark_probability_threshold,
        edge_columns,
    ].to_numpy(),
))
true_edges = set(map(tuple, ground_truth[edge_columns].to_numpy()))

print("\n--- Resultados do Ensemble (Probabilístico) ---")
print(synth_summary[["source", "target", "lag", "edge_probability"]])

print("\n--- Diagnóstico das Arestas ---")
print("Verdadeiros positivos:", sorted(predicted_edges & true_edges))
print("Falsos positivos:", sorted(predicted_edges - true_edges))
print("Falsos negativos:", sorted(true_edges - predicted_edges))

print("\n--- Métricas Estruturais (Benchmark) ---")
for k, v in metrics.items():
    print(f"{k}: {v:.2f}" if isinstance(v, float) else f"{k}: {v}")

--- Ground Truth Esperado ---
  source target  lag
0      X      Y    2
1      Y      Z    1
2      X      Z    1



--- Resultados do Ensemble (Probabilístico) ---
  source target  lag  edge_probability
0      Y      Z    1               1.0
1      X      Y    2               1.0
2      X      Z    1               1.0

--- Diagnóstico das Arestas ---
Verdadeiros positivos: [('X', 'Y', 2), ('X', 'Z', 1), ('Y', 'Z', 1)]
Falsos positivos: []
Falsos negativos: []

--- Métricas Estruturais (Benchmark) ---
true_positives: 3
false_positives: 0
false_negatives: 0
reversed_edges: 0
precision: 1.00
recall: 1.00
f1_score: 1.00
structural_hamming_distance: 0


## Teste de robustez sob mudança no regime de ruído

Compara a recuperação da estrutura causal no benchmark sintético limpo e após a introdução de ruído severo.

In [14]:
from causal_discovery.benchmark import inject_noise_regime_change

# 4. Mudança no Regime de Ruído (Teste de Robustez)
print("--- Teste de Robustez à Mudança no Regime de Ruído ---")

# Aumenta o ruído a partir da posição 250 sem alterar a estrutura causal geradora
df_noisy = inject_noise_regime_change(df_synth, index_change=250, noise_multiplier=3.0)

# Roda a mesma seleção robusta na série corrompida
noisy_selection = select_robust_ensemble_combination(
    df_noisy,
    benchmark_methods,
    method_kwargs=benchmark_method_kwargs,
    method_weights=all_method_weights,
    min_methods=benchmark_min_methods,
    max_methods=benchmark_max_methods,
    min_votes=benchmark_min_votes,
    n_bootstrap=benchmark_n_bootstrap,
    selection_probability_threshold=benchmark_selection_threshold,
    random_state=42,
)

noisy_summary = noisy_selection["best_evaluation"]["probabilistic_summary"]
noisy_summary = noisy_summary.loc[
    noisy_summary["source"].ne(noisy_summary["target"])
].reset_index(drop=True)
noisy_metrics = compute_structural_metrics(
    noisy_summary,
    ground_truth,
    prob_threshold=benchmark_probability_threshold,
)

print("Métricas originais (Limpo):", {k: round(v, 2) if isinstance(v, float) else v for k,v in metrics.items() if k in ['f1_score', 'structural_hamming_distance']})
print("Métricas com Ruído Severo:", {k: round(v, 2) if isinstance(v, float) else v for k,v in noisy_metrics.items() if k in ['f1_score', 'structural_hamming_distance']})

--- Teste de Robustez à Mudança no Regime de Ruído ---


Métricas originais (Limpo): {'f1_score': 1.0, 'structural_hamming_distance': 0}
Métricas com Ruído Severo: {'f1_score': 0.5, 'structural_hamming_distance': 2}


# Selecao de algoritmos via causal_algorithms_atlas (perfil + RAG)

Esta secao conecta a base de conhecimento verificada em `causal_algorithms_atlas/`
ao pipeline de ensemble deste notebook. O fluxo:

1. **Perfilar o dataset** (`dataset_profile.profile_dataset`): estacionariedade
   (teste ADF) e linearidade -- comparando erro de previsao FORA DA AMOSTRA
   (janela expansiva, ridge regression) entre um modelo linear e um com termos
   quadraticos/cubicos sobre o lag 1 de todas as variaveis. Isso mede se a nao
   linearidade e grande o bastante para importar NA PRATICA, nao apenas se e
   estatisticamente detectavel (um teste de significancia pura, como usamos
   numa versao anterior, rejeita linearidade para qualquer desvio assim que ha
   dados suficientes -- inutil como filtro em dado real).
2. **Recomendar metodos** (`ensemble_advisor.recommend_framework_methods`): cruza
   o perfil com as premissas **estruturadas** e verificadas de cada ficha do atlas.
   Cada inclusao/exclusao vem com o motivo explicito.
3. **Selecionar o ensemble** (`causal_discovery.select_robust_ensemble_combination`,
   ja existente no framework): escolhe a melhor combinacao usando **apenas
   metricas cegas** -- estabilidade sob bootstrap, confianca, densidade. O ground
   truth **nunca** entra nessa escolha.
4. **Avaliar depois** (`experiment_runner.run_experiment`): so agora, com a
   composicao ja decidida, o ground truth (quando existe) e consultado para
   reportar precision/recall/F1 -- e comparar com o melhor metodo individual,
   tambem escolhido sem olhar o gabarito.

Essa ordem (selecionar cego, avaliar depois) e o que torna valida qualquer
afirmacao de que "o ensemble e pelo menos tao bom quanto o melhor metodo sozinho":
a composicao nao foi ajustada olhando a resposta certa.

Todo o historico de execucoes desta secao fica registrado em
`.local/ensemble_rag_experiments/history.jsonl` (nao versionado).

In [ ]:
from causal_discovery import load_time_series_dataset
from causal_algorithms_atlas.dataset_profile import profile_dataset
from causal_algorithms_atlas.ensemble_advisor import recommend_framework_methods

# Perfil do dataset sintetico linear (ground truth conhecido: X1->Y, X3->Y, X4->X1)
toy_linear = load_time_series_dataset(
    "datasets/synthetic_causal/toy_a_linear.csv",
    ground_truth_path="datasets/synthetic_causal/toy_a_linear_gt.csv",
)
profile = profile_dataset(toy_linear.data)
print(profile.to_query_text())
print()

recommendations = recommend_framework_methods(profile)
print(f"{'metodo':<20} {'incluido':<10} motivo")
for rec in recommendations:
    print(f"{rec.framework_method_name:<20} {str(rec.included):<10} {' '.join(rec.reasons)}")


Dataset com 5 variaveis e 20000 observacoes. 100% das series testadas sao estacionarias (teste ADF, alfa=0.05). 100% das series testadas tem relacao com o lag 1 de si mesma e das demais variaveis aproximadamente lineares (comparando erro de previsao fora da amostra entre um modelo linear e um modelo com termos quadraticos). Confundidores latentes nao sao verificaveis apenas a partir dos dados observados (limite de identificabilidade, nao uma medida deste perfil). Quais algoritmos de causal discovery sao mais adequados para este perfil?

metodo               incluido   motivo
ClassicalGranger     True       Exige estacionariedade; 100% das series testadas passaram no ADF. Assume relacoes lineares; em 100% das series testadas, permitir termos nao lineares nao reduziu o erro de previsao fora da amostra o suficiente para importar na pratica.
DYNOTEARS            True       Assume relacoes lineares; em 100% das series testadas, permitir termos nao lineares nao reduziu o erro de previsao for

## toy_a_linear: fluxo completo, dataset facil (efeito teto)

Dataset sintetico linear e estacionario com apenas 3 relacoes causais verdadeiras
(`X1->Y`, `X3->Y`, `X4->X1`). Os 8 metodos do framework passam no filtro de perfil
(todas as premissas de linearidade/estacionariedade se aplicam). Ensemble e melhor
metodo sozinho empatam em F1=1.0 -- **este dataset e facil demais para diferenciar
ensemble de metodo unico** (efeito teto, nao falha do metodo).

In [ ]:
from causal_algorithms_atlas.experiment_runner import run_experiment

# Roda leve (amostra + poucos bootstraps) para reexecucao rapida deste notebook.
# A composicao do ensemble e escolhida SOMENTE por metricas cegas (estabilidade sob
# bootstrap, confianca, densidade) -- o ground truth so entra DEPOIS, para reportar
# o desempenho. Isso e o que garante que "o ensemble bate o melhor metodo sozinho"
# nao seja uma afirmacao obtida "roubando" (escolhendo a composicao que ja sabemos
# que acerta o gabarito).
HISTORY_PATH = ".local/ensemble_rag_experiments/history.jsonl"

result = run_experiment(
    toy_linear.data.iloc[:500].reset_index(drop=True),
    ground_truth=toy_linear.ground_truth,
    dataset_name="toy_a_linear_notebook_run",
    n_bootstrap=3,
    max_methods=3,
    random_state=42,
    history_path=HISTORY_PATH,
)

print("Metodos candidatos (compativeis com o perfil):", result["candidate_methods"])
print()
print(
    f"Melhor combinacao (escolhida SEM olhar o ground truth): "
    f"{result['best_combination_methods']} "
    f"(performance_score={result['best_combination_performance_score']:.4f})"
)
print(
    f"Melhor metodo sozinho (escolhido SEM olhar o ground truth): "
    f"{result['best_single_method']} "
    f"(performance_score={result['best_single_performance_score']:.4f})"
)
print()
print("--- Avaliacao pos-hoc contra o ground truth (so agora consultado) ---")
print("Combinacao:", result["best_combination_metrics_post_hoc"])
print("Melhor metodo sozinho:", result["best_single_metrics_post_hoc"])
print()
print("Ensemble >= melhor metodo sozinho (F1)?", result["ensemble_beats_best_single_f1"])


Metodos candidatos (compativeis com o perfil): ['ClassicalGranger', 'DYNOTEARS', 'FCI', 'GES', 'LPCMCI', 'NeuralGrangercMLP', 'PCMCI', 'VARLiNGAM']

Melhor combinacao (escolhida SEM olhar o ground truth): ['ClassicalGranger', 'FCI', 'PCMCI'] (performance_score=0.8490)
Melhor metodo sozinho (escolhido SEM olhar o ground truth): PCMCI (performance_score=0.8026)

--- Avaliacao pos-hoc contra o ground truth (so agora consultado) ---
Combinacao: {'precision': 1.0, 'recall': 1.0, 'f1_score': 1.0, 'true_positives': 3, 'false_positives': 0, 'false_negatives': 0}
Melhor metodo sozinho: {'precision': 1.0, 'recall': 1.0, 'f1_score': 1.0, 'true_positives': 3, 'false_positives': 0, 'false_negatives': 0}

Ensemble >= melhor metodo sozinho (F1)? True


## toy_b_nonlinear: aqui o ensemble bate o melhor metodo sozinho

`toy_b_nonlinear` tem as mesmas 3 relacoes causais de `toy_a_linear`, mas geradas
com `tanh(.)` em vez de lineares. O perfilador agora detecta corretamente que `Y`
e `X1` (as duas variaveis que recebem uma entrada transformada por tanh) tem
nao linearidade pratica (ganho preditivo fora da amostra de ~1%, acima do limiar
calibrado), mas isso nao muda a decisao agregada do dataset (60% das series ainda
sao predominantemente lineares), entao os 8 metodos continuam candidatos.

O resultado: o **melhor metodo sozinho escolhido as cegas foi GES** (F1=0.80,
perdeu a aresta `X4->X1`), enquanto a **combinacao escolhida as cegas** (
`ClassicalGranger + DYNOTEARS + PCMCI`) **acertou as 3 arestas (F1=1.00)**. Esta
e a primeira evidencia neste notebook de que o ensemble supera o melhor metodo
individual sem nenhum ajuste que olhe o ground truth durante a selecao.

In [ ]:
from causal_discovery import load_daily_delhi_climate
from causal_algorithms_atlas.experiment_runner import (
    InsufficientCandidatesError,
    run_experiment,
)

toy_nonlinear = load_time_series_dataset(
    "datasets/synthetic_causal/toy_b_nonlinear.csv",
    ground_truth_path="datasets/synthetic_causal/toy_b_nonlinear_gt.csv",
)
delhi = load_daily_delhi_climate("DailyDelhiClimateTrain.csv")

for name, data, gt in [
    ("toy_b_nonlinear", toy_nonlinear.data, toy_nonlinear.ground_truth),
    ("daily_delhi_climate", delhi, None),
]:
    print(f"=== {name} ===")
    try:
        result = run_experiment(
            data,
            ground_truth=gt,
            dataset_name=f"{name}_notebook_run",
            n_bootstrap=3,
            max_methods=3,
            random_state=42,
            history_path=".local/ensemble_rag_experiments/history.jsonl",
        )
    except InsufficientCandidatesError as exc:
        print("Ensemble nao formado --", exc)
        print()
        continue

    print("Candidatos compativeis:", result["candidate_methods"])
    print(
        f"Melhor combinacao (cega ao gt): {result['best_combination_methods']} "
        f"(performance_score={result['best_combination_performance_score']:.4f})"
    )
    print(
        f"Melhor metodo sozinho (cego ao gt): {result['best_single_method']} "
        f"(performance_score={result['best_single_performance_score']:.4f})"
    )
    if result["best_combination_metrics_post_hoc"] is not None:
        print("F1 pos-hoc combinacao:", result["best_combination_metrics_post_hoc"])
        print("F1 pos-hoc melhor sozinho:", result["best_single_metrics_post_hoc"])
    else:
        print("(sem ground truth disponivel para este dataset -- so metricas cegas)")
    print()


=== toy_b_nonlinear ===
Candidatos compativeis: ['ClassicalGranger', 'DYNOTEARS', 'FCI', 'GES', 'LPCMCI', 'NeuralGrangercMLP', 'PCMCI', 'VARLiNGAM']
Melhor combinacao (cega ao gt): ['ClassicalGranger', 'DYNOTEARS', 'PCMCI'] (performance_score=0.8490)
Melhor metodo sozinho (cego ao gt): GES (performance_score=0.7998)
F1 pos-hoc combinacao: {'precision': 1.0, 'recall': 1.0, 'f1_score': 1.0, 'true_positives': 3, 'false_positives': 0, 'false_negatives': 0}
F1 pos-hoc melhor sozinho: {'precision': 1.0, 'recall': 0.6666666666666666, 'f1_score': 0.8, 'true_positives': 2, 'false_positives': 0, 'false_negatives': 1}

=== daily_delhi_climate ===
Ensemble nao formado -- Apenas 1 metodo(s) do framework sao compativeis com o perfil deste dataset (daily_delhi_climate_notebook_run): ['NeuralGrangercMLP']. Um ensemble precisa de pelo menos 2 metodos concordando (min_votes>=2). Metodos excluidos e motivo: ['ClassicalGranger (Exige estacionariedade; 75% das series testadas passaram no ADF.; Assume relac

## O limite real: dataset real onde so 1 metodo sobra

Tentamos o mesmo fluxo no dataset real do projeto
(`DailyDelhiClimateTrain.csv`, sem ground truth). Aqui, `meantemp`, `humidity` e
`meanpressure` mostram nao linearidade pratica (ganho preditivo de 3-6% fora da
amostra) grande o bastante para dominar a decisao agregada (75% das series nao
lineares) -- **so `NeuralGrangercMLP` sobrevive ao filtro**, e sem um segundo
metodo concordando, nenhum ensemble pode ser formado (`InsufficientCandidatesError`).

Isso e fisicamente plausivel, nao um artefato do filtro: a relacao
temperatura-umidade tem uma base fisica genuinamente nao linear (equacao de
Clausius-Clapeyron para pressao de vapor de saturacao). O resultado exposto aqui
nao e um bug do perfilador -- e uma lacuna real de cobertura do framework: dos 8
metodos verificados no atlas, apenas 1 (`NeuralGrangercMLP`) tolera nao
linearidade. Os outros 7 candidatos verificados na literatura mas nao
implementados (`transfer_entropy`, `tcdf`, `ccm`, `timino`, ver
`causal_algorithms_atlas/algorithms/`) seriam os proximos candidatos naturais a
implementar se este tipo de dataset for uma prioridade.

## Historico desta secao (o que mudou e por que)

O criterio de linearidade do perfilador passou por duas versoes nesta mesma
investigacao, documentadas nos commits de `causal_algorithms_atlas/dataset_profile.py`:

1. **V1 -- teste de significancia (RESET de Ramsey):** testava "existe nao
   linearidade detectavel". Problema real: com dados suficientes, qualquer
   desvio de linearidade vira "estatisticamente significativo", mesmo sendo
   irrelevante na pratica. Isso deixava so 1 dos 8 metodos como candidato tanto
   em `toy_b_nonlinear` quanto no dataset real.
2. **V2 -- tamanho de efeito (esta versao):** testa "a nao linearidade reduz o
   erro de previsao fora da amostra o suficiente para importar". Corrigiu dois
   bugs numericos reais no caminho (padronizar antes de elevar a potencia;
   winsorizar outliers de sensor reais em `DailyDelhiClimateTrain.csv`, que sem
   isso geravam "ganhos preditivos" na casa dos bilhoes). O limiar
   (`_NONLINEARITY_EFFECT_SIZE_THRESHOLD = 0.005`) foi calibrado empiricamente
   comparando o "chao de ruido" de relacoes genuinamente lineares (`toy_a_linear`,
   <0.001) com a nao linearidade real e conhecida de `toy_b_nonlinear` (Y e X1,
   0.009-0.012) -- nao e um valor convencional como um alfa de significancia, e
   pode precisar de revisao se novos datasets mostrarem uma zona cinzenta
   diferente.

A exclusao binaria de metodos incompativeis foi mantida como pedido -- o que
mudou foi so o criterio usado para decidir a exclusao.